### ISSUE TITLE
Modernize shutdown

### ISSUE THREAD

**Issue by avikivity** [scylladb/scylla#2737]
Currently, startup/shutdown is a mess of hacks.

Here's a plan to modernize it:

- hook SIGINT/SIGTERM to a boost::signals2::signal that broadcasts the shutdown state
- every component with long-running tasks (e.g. repair, compaction) will listen to this signal and stop activity when it is signalled. Typically by setting some flag which is polled in the main loop of the task.
- an additional listener will be a promise::set_value() call to a promise named stopping
- main will be adjusted to look like
```
    start1
    start2
    start3
    stopping.get();
    shut1
    shut2
    shut3
```
 - we can make use of defer():
```
    foo.start().get();
    defer([&] { foo.stop().get(); });
    ...
    stopping.get()
```
 - make use of `seastar::sharded`'s ability to pass other `sharded` services for explicit dependencies instead of global scattered all over the code.

**Comment by avikivity** [comment]
I believe this work was completed. @xemul are you aware of counterexamples?

**Comment by xemul** [comment]
Yes, and all of them are hanging aroung the `drain_on_shutdown`.

**Comment by avikivity** [comment]
:(

**Comment by dani-tweig** [comment]
@xemul , what is the status of this one? (We are trying to understand the content of 6.1, and if we have release blockers that we should wait for)

**Comment by xemul** [comment]
The services that were "out of order" when I picked this issue up are now all fixed. However, wile I was fixing those, there appeared group0 and messed up the whole main() :( I didn't touch it yet, because it was is active development

And #19682 to polish things up

**Comment by kbr-scylla** [comment]
But group 0 related services are following the modern practice of "start, defer shutdown", no?

**Comment by xemul** [comment]
No

```
service::raft_group0_client group0_client{...}; // created non-sharded
service::raft_group0_client group0_service{..., group0_client, ...}; // created non-sharded
...
mm.start(..., std::ref(group0_client), ...); // sharded migration manager starts
...
qp.start(..., std::ref(group0_client), ...); // sharded query processor starts
...
group0_client.init().get(); // sort of .start(), but out of order
... // more service start and initialize here
group0_service.start().get(); // now it starts _way_ after it was created
ss.local().set_group0(group0_service); // storage service gets it dependency after it started, but it should have got one via constructor
```

also raft group registry is started-and-stopped like

```
sharded<raft_group_registry> raft_gr;
...
raft_gr.start().get(); // sharded start, that's fine
... // more services that depend on raft_gr start and stop here
// XXX: stop_raft has to happen before query_processor and migration_manager
// is stopped, since some groups keep using the query
// processor until are stopped inside stop_raft.
auto stop_raft = defer([] { raft_gr.stop().get(); }); // deferred stop of raft_gr, should immediately follow its start
... // more services start and stop here
raft_gr.invoke_on_all(&service::raft_group_registry::start).get(); // start of individual instances of raft_gr, should follow its sharded<>::start()
```

**Comment by xemul** [comment]
Ah, also tracing fix wasn't merged, so it still needs care (#14156)

**Comment by dani-tweig** [comment]
@xemul, after all the latest commits, can you summarize what else needs to be done here?
or can we close it?

**Comment by xemul** [comment]
There's still a mess in start-drain-stop of
* tracing (however, there's a fix for that #14156)
* raft_group_registry, group0_client and raft_group0_client (it's been changed a lot while developing, so I didn't touch it, hopefully it's stable enough already)
